File: Backend/api/v1/tools/map.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import List, Optional
import math

router = APIRouter()


class RoutePoint(BaseModel):
    lat: float
    lng: float
    name: Optional[str] = None


class RouteRequest(BaseModel):
    origin: RoutePoint
    destination: RoutePoint
    waypoints: Optional[List[RoutePoint]] = []


class RouteStep(BaseModel):
    instruction: str
    distance: float  # in meters
    duration: float  # in seconds
    start_location: RoutePoint
    end_location: RoutePoint


class RouteResponse(BaseModel):
    distance: float  # in meters
    duration: float  # in seconds
    steps: List[RouteStep]
    polyline: List[RoutePoint]  # route coordinates


def calculate_distance(lat1: float, lng1: float, lat2: float, lng2: float) -> float:
    """
    Calculate the great circle distance between two points 
    on the earth (specified in decimal degrees)
    Returns distance in meters
    """
    # Convert decimal degrees to radians
    lat1, lng1, lat2, lng2 = map(math.radians, [lat1, lng1, lat2, lng2])

    # Haversine formula
    dlat = lat2 - lat1
    dlon = lng2 - lng1
    a = math.sin(dlat/2)**2 + math.cos(lat1) * \
        math.cos(lat2) * math.sin(dlon/2)**2
    c = 2 * math.asin(math.sqrt(a))
    r = 6371000  # Radius of earth in meters
    return c * r


def calculate_bearing(lat1: float, lng1: float, lat2: float, lng2: float) -> float:
    """
    Calculate the initial bearing between two points
    """
    lat1, lng1, lat2, lng2 = map(math.radians, [lat1, lng1, lat2, lng2])
    dLng = lng2 - lng1

    y = math.sin(dLng) * math.cos(lat2)
    x = math.cos(lat1) * math.sin(lat2) - math.sin(lat1) * \
        math.cos(lat2) * math.cos(dLng)
    brng = math.atan2(y, x)
    brng = math.degrees(brng)
    return (brng + 360) % 360


def generate_mock_route(origin: RoutePoint, destination: RoutePoint) -> RouteResponse:
    """
    Generate a mock route between origin and destination
    In a real implementation, this would call a mapping service like AMap or Google Maps API
    """
    distance = calculate_distance(
        origin.lat, origin.lng, destination.lat, destination.lng)

    # Estimate duration based on distance (assuming average speed of 30 km/h)
    avg_speed_ms = 30000 / 3600  # meters per second
    duration = distance / avg_speed_ms

    # Generate intermediate points for the route
    steps = []
    polyline = []

    # Add start point
    polyline.append(RoutePoint(lat=origin.lat, lng=origin.lng))

    # Calculate bearing from origin to destination
    bearing = calculate_bearing(
        origin.lat, origin.lng, destination.lat, destination.lng)

    # Generate intermediate points along the route
    # Between 2-20 points based on distance
    num_points = max(2, min(int(distance / 1000) + 1, 20))
    for i in range(1, num_points):
        fraction = i / num_points

        # Calculate intermediate point using bearing and distance
        lat_diff = (destination.lat - origin.lat) * fraction
        lng_diff = (destination.lng - origin.lng) * fraction

        intermediate_lat = origin.lat + lat_diff
        intermediate_lng = origin.lng + lng_diff

        polyline.append(RoutePoint(lat=intermediate_lat, lng=intermediate_lng))

        # Add step instruction
        steps.append(RouteStep(
            instruction=f"沿当前方向行驶约 {int(distance / num_points)} 米",
            distance=distance / num_points,
            duration=duration / num_points,
            start_location=polyline[-2],
            end_location=polyline[-1]
        ))

    # Add final destination
    polyline.append(RoutePoint(lat=destination.lat, lng=destination.lng))

    # Add final step
    steps.append(RouteStep(
        instruction="到达目的地",
        distance=calculate_distance(
            polyline[-2].lat, polyline[-2].lng, destination.lat, destination.lng),
        duration=duration / num_points,
        start_location=polyline[-2],
        end_location=polyline[-1]
    ))

    return RouteResponse(
        distance=distance,
        duration=duration,
        steps=steps,
        polyline=polyline
    )


@router.post("/route", response_model=RouteResponse, tags=["map"])
async def get_route(request: RouteRequest):
    """
    Get route between origin and destination
    """
    try:
        route = generate_mock_route(request.origin, request.destination)
        return route
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.get("/route/test", tags=["map"])
async def test_route():
    """
    Test endpoint for route functionality
    """
    return {"message": "Route API is working"}

File: Backup/tools_project_rust/src/services/tools/map.rs
use crate::models::{RoutePoint, RouteRequest, RouteResponse, RouteStep};

const EARTH_RADIUS: f64 = 6371000.0;
const AVG_SPEED_MS: f64 = 30000.0 / 3600.0;

fn to_radians(deg: f64) -> f64 {
    deg * std::f64::consts::PI / 180.0
}

fn calculate_distance(lat1: f64, lng1: f64, lat2: f64, lng2: f64) -> f64 {
    let lat1 = to_radians(lat1);
    let lng1 = to_radians(lng1);
    let lat2 = to_radians(lat2);
    let lng2 = to_radians(lng2);

    let dlat = lat2 - lat1;
    let dlon = lng2 - lng1;
    let a = (dlat / 2.0).sin().powi(2)
        + lat1.cos() * lat2.cos() * (dlon / 2.0).sin().powi(2);
    let c = 2.0 * a.sqrt().asin();
    c * EARTH_RADIUS
}

pub fn generate_mock_route(request: &RouteRequest) -> RouteResponse {
    let origin = &request.origin;
    let destination = &request.destination;
    let distance = calculate_distance(origin.lat, origin.lng, destination.lat, destination.lng);
    let duration = distance / AVG_SPEED_MS;

    let mut polyline = vec![RoutePoint {
        lat: origin.lat,
        lng: origin.lng,
        name: origin.name.clone(),
    }];
    let mut steps = vec![];

    let num_points = (distance / 1000.0) as usize + 1;
    let num_points = num_points.clamp(2, 20);

    for i in 1..num_points {
        let fraction = i as f64 / num_points as f64;
        let lat = origin.lat + (destination.lat - origin.lat) * fraction;
        let lng = origin.lng + (destination.lng - origin.lng) * fraction;
        let point = RoutePoint { lat, lng, name: None };
        let prev = polyline.last().unwrap().clone();
        steps.push(RouteStep {
            instruction: format!("沿当前方向行驶约 {} 米", (distance / num_points as f64) as i64),
            distance: distance / num_points as f64,
            duration: duration / num_points as f64,
            start_location: prev,
            end_location: point.clone(),
        });
        polyline.push(point);
    }

    let final_distance = calculate_distance(
        polyline.last().unwrap().lat,
        polyline.last().unwrap().lng,
        destination.lat,
        destination.lng,
    );
    steps.push(RouteStep {
        instruction: "到达目的地".to_string(),
        distance: final_distance,
        duration: duration / num_points as f64,
        start_location: polyline.last().unwrap().clone(),
        end_location: RoutePoint {
            lat: destination.lat,
            lng: destination.lng,
            name: destination.name.clone(),
        },
    });
    polyline.push(RoutePoint {
        lat: destination.lat,
        lng: destination.lng,
        name: destination.name.clone(),
    });

    RouteResponse {
        distance,
        duration,
        steps,
        polyline,
    }
}
